openmdao/solvers/petsc_ksp.py

```diff
@@ -0,0 +1,198 @@
+            interest. Each array contains the right-hand side for the linear
+            solve.
+
+        system : `System`
+            Parent `System` object.
+
+        mode : string
+            Derivative mode, can be 'fwd' or 'rev'.
+
+        Returns
+        -------
+        dict of ndarray : Solution vectors
+        """
+        options = self.options
+        self.mode = mode
+
+        self.ksp.setTolerances(max_it=options['maxiter'],
+                               atol=options['atol'],
+                               rtol=options['rtol'])
+
+        unknowns_mat = {}
+        for voi, rhs in iteritems(rhs_mat):
+
+            sol_vec = np.zeros(rhs.shape)
+            # Set these in the system
+            if trace:
+                debug("creating sol_buf petsc vec for voi", voi)
+            self.sol_buf_petsc = PETSc.Vec().createWithArray(sol_vec,
+                                                             comm=system.comm)
+            if trace:
+                debug("creating rhs_buf petsc vec for voi", voi)
+            self.rhs_buf_petsc = PETSc.Vec().createWithArray(rhs,
+                                                             comm=system.comm)
+
+            # Petsc can only handle one right-hand-side at a time for now
+            self.voi = voi
+            self.system = system
+            self.ksp.solve(self.rhs_buf_petsc, self.sol_buf_petsc)
+            self.system = None
+
+            unknowns_mat[voi] = sol_vec
+
+            #print system.name, 'Linear solution vec', d_unknowns
+
+        self.system = None
+        return unknowns_mat
+
+    def mult(self, mat, arg, result):
+        """ KSP Callback: applies Jacobian matrix. Mode is determined by the
+        system.
+
+        Args
+        ----
+        arg : PetSC Vector
+            Incoming vector
+
+        result : PetSC Vector
+            Empty array into which we place the matrix-vector product.
+        """
+
+        system = self.system
+        mode = self.mode
+
+        voi = self.voi
+        if mode == 'fwd':
+            sol_vec, rhs_vec = system.dumat[voi], system.drmat[voi]
+        else:
+            sol_vec, rhs_vec = system.drmat[voi], system.dumat[voi]
+
+        # Set incoming vector
+        sol_vec.vec[:] = arg.array[:]
+
+        # Start with a clean slate
+        rhs_vec.vec[:] = 0.0
+        system.clear_dparams()
+
+        system.apply_linear(mode, ls_inputs=self.system._ls_inputs, vois=[voi])
+
+        result.array[:] = rhs_vec.vec[:]
+        #print("arg", arg.array)
+        #print("result", result.array)
+
+    def apply(self, mat, sol_vec, rhs_vec):
+        """ Applies preconditioner
+
+        Args
+        ----
+        sol_vec : PetSC Vector
+            Incoming vector
+
+        rhs_vec : PetSC Vector
+            Empty vector into which we return the preconditioned sol_vec
+        """
+
+        # TODO - Preconditioning is not supported yet, so mimic an Identity
```

Comment:
why is preconditioning not supported? We have the apply_linear method in there, no?